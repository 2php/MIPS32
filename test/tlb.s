# tlb
.set noreorder
start:
  .text 0
  jal init
  nop
  
  la $s0, 0xbfd003f8
  la $s1, 0xbfd003fc
  
  lw $a0, 0x40000070($0)
  jal write_com
  nop
  
  break

init:
  
  # set exception return base (EBase)
  la $t1, 0x80000000 
  mfc0 $t0, $15
  or   $t0, $t1
  mtc0 $t0, $15
  
  # enable exception
  mfc0 $t0, $12
  or   $t0,  0b00000000000000000000000000000001
  and  $t0,~(0b00000000000000000000000000000010)
  mtc0 $t0, $12
  
  jr $ra
  nop

write_com:
  li $t1, 1
  loop:
  lw $t0, 0($s1)
  andi $t0, $t0, 1
  bne $t0, $t1, loop
  sw $a0, 0($s0)
  jr $ra
  nop
  
data:
  .word 0x00000012
  
# ex_handler @ 0x80000180
.set noreorder
start:
  .text 0
  
  la $k0, 0x00000000
  mtc0 $k0, $0
  
  la $k0, 0x00000002
  mtc0 $k0, $2
  
  la $k0, 0x40000000
  mtc0 $k0, $10
  
  tlbwi
  
  eret
 