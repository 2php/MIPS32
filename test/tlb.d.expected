booting

fetch instr @ 0x80000000 (10000000000000000000000000000000)
MMU[0x80000000] :  0x00000000
Mem[0x00000000] :  0x0c00000b (00001100000000000000000000001011)
000011 | 00000000000000000000001011
JAL      instr_index:0x000000b
R[31] <= 0x80000008 (10000000000000000000000000001000)

fetch instr @ 0x8000002c (10000000000000000000000000101100)
MMU[0x8000002c] :  0x0000002c
Mem[0x0000002c] :  0x3c098000 (00111100000010011000000000000000)
001111 | 00000 | 01001 | 1000000000000000
LUI      rs: 0   rt: 9   imm:0x8000
R[ 9] <= 0x80000000 (10000000000000000000000000000000)

fetch instr @ 0x80000030 (10000000000000000000000000110000)
MMU[0x80000030] :  0x00000030
Mem[0x00000030] :  0x40087800 (01000000000010000111100000000000)
010000 | 00000 | 01000 | 01111 | 00000000 | 000
COP0     mfc0    rt: 8   EBase   
R[EBase   ] :  0x80000000 (10000000000000000000000000000000)
R[ 8] <= 0x80000000 (10000000000000000000000000000000)

fetch instr @ 0x80000034 (10000000000000000000000000110100)
MMU[0x80000034] :  0x00000034
Mem[0x00000034] :  0x01094025 (00000001000010010100000000100101)
000000 | 01000 | 01001 | 01000 | 00000 | 100101
SPECIAL  rs: 8   rt: 9   rd: 8   sa: 0   OR     
R[ 8] :  0x80000000 (10000000000000000000000000000000)
R[ 9] :  0x80000000 (10000000000000000000000000000000)
alu_a <= 0x80000000 (10000000000000000000000000000000)
alu_b <= 0x80000000 (10000000000000000000000000000000)
alu_r :  0x80000000 (10000000000000000000000000000000)
R[ 8] <= 0x80000000 (10000000000000000000000000000000)

fetch instr @ 0x80000038 (10000000000000000000000000111000)
MMU[0x80000038] :  0x00000038
Mem[0x00000038] :  0x40887800 (01000000100010000111100000000000)
010000 | 00100 | 01000 | 01111 | 00000000 | 000
COP0     mtc0    rt: 8   EBase   
R[ 8] :  0x80000000 (10000000000000000000000000000000)
R[EBase   ] <= 0x80000000 (10000000000000000000000000000000)

fetch instr @ 0x8000003c (10000000000000000000000000111100)
MMU[0x8000003c] :  0x0000003c
Mem[0x0000003c] :  0x40086000 (01000000000010000110000000000000)
010000 | 00000 | 01000 | 01100 | 00000000 | 000
COP0     mfc0    rt: 8   SR      
R[SR      ] :  0x10000002 (00010000000000000000000000000010)
R[ 8] <= 0x10000002 (00010000000000000000000000000010)

fetch instr @ 0x80000040 (10000000000000000000000001000000)
MMU[0x80000040] :  0x00000040
Mem[0x00000040] :  0x35080001 (00110101000010000000000000000001)
001101 | 01000 | 01000 | 0000000000000001
ORI      rs: 8   rt: 8   imm:0x0001
R[ 8] :  0x10000002 (00010000000000000000000000000010)
alu_a <= 0x10000002 (00010000000000000000000000000010)
alu_b <= 0x00000001 (00000000000000000000000000000001)
alu_r :  0x10000003 (00010000000000000000000000000011)
R[ 8] <= 0x10000003 (00010000000000000000000000000011)

fetch instr @ 0x80000044 (10000000000000000000000001000100)
MMU[0x80000044] :  0x00000044
Mem[0x00000044] :  0x2401fffd (00100100000000011111111111111101)
001001 | 00000 | 00001 | 1111111111111101
ADDIU    rs: 0   rt: 1   imm:0xfffd
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0xfffffffd (11111111111111111111111111111101)
alu_r :  0xfffffffd (11111111111111111111111111111101)
R[ 1] <= 0xfffffffd (11111111111111111111111111111101)

fetch instr @ 0x80000048 (10000000000000000000000001001000)
MMU[0x80000048] :  0x00000048
Mem[0x00000048] :  0x01014024 (00000001000000010100000000100100)
000000 | 01000 | 00001 | 01000 | 00000 | 100100
SPECIAL  rs: 8   rt: 1   rd: 8   sa: 0   AND    
R[ 8] :  0x10000003 (00010000000000000000000000000011)
R[ 1] :  0xfffffffd (11111111111111111111111111111101)
alu_a <= 0x10000003 (00010000000000000000000000000011)
alu_b <= 0xfffffffd (11111111111111111111111111111101)
alu_r :  0x10000001 (00010000000000000000000000000001)
R[ 8] <= 0x10000001 (00010000000000000000000000000001)

fetch instr @ 0x8000004c (10000000000000000000000001001100)
MMU[0x8000004c] :  0x0000004c
Mem[0x0000004c] :  0x40886000 (01000000100010000110000000000000)
010000 | 00100 | 01000 | 01100 | 00000000 | 000
COP0     mtc0    rt: 8   SR      
R[ 8] :  0x10000001 (00010000000000000000000000000001)
R[SR      ] <= 0x10000001 (00010000000000000000000000000001)

fetch instr @ 0x80000050 (10000000000000000000000001010000)
MMU[0x80000050] :  0x00000050
Mem[0x00000050] :  0x03e00008 (00000011111000000000000000001000)
000000 | 11111 | 00000 | 00000 | 00000 | 001000
SPECIAL  rs:31   rt: 0   rd: 0   sa: 0   JR     
R[31] :  0x80000008 (10000000000000000000000000001000)

fetch instr @ 0x80000008 (10000000000000000000000000001000)
MMU[0x80000008] :  0x00000008
Mem[0x00000008] :  0x3c10bfd0 (00111100000100001011111111010000)
001111 | 00000 | 10000 | 1011111111010000
LUI      rs: 0   rt:16   imm:0xbfd0
R[16] <= 0xbfd00000 (10111111110100000000000000000000)

fetch instr @ 0x8000000c (10000000000000000000000000001100)
MMU[0x8000000c] :  0x0000000c
Mem[0x0000000c] :  0x361003f8 (00110110000100000000001111111000)
001101 | 10000 | 10000 | 0000001111111000
ORI      rs:16   rt:16   imm:0x03f8
R[16] :  0xbfd00000 (10111111110100000000000000000000)
alu_a <= 0xbfd00000 (10111111110100000000000000000000)
alu_b <= 0x000003f8 (00000000000000000000001111111000)
alu_r :  0xbfd003f8 (10111111110100000000001111111000)
R[16] <= 0xbfd003f8 (10111111110100000000001111111000)

fetch instr @ 0x80000010 (10000000000000000000000000010000)
MMU[0x80000010] :  0x00000010
Mem[0x00000010] :  0x3c11bfd0 (00111100000100011011111111010000)
001111 | 00000 | 10001 | 1011111111010000
LUI      rs: 0   rt:17   imm:0xbfd0
R[17] <= 0xbfd00000 (10111111110100000000000000000000)

fetch instr @ 0x80000014 (10000000000000000000000000010100)
MMU[0x80000014] :  0x00000014
Mem[0x00000014] :  0x363103fc (00110110001100010000001111111100)
001101 | 10001 | 10001 | 0000001111111100
ORI      rs:17   rt:17   imm:0x03fc
R[17] :  0xbfd00000 (10111111110100000000000000000000)
alu_a <= 0xbfd00000 (10111111110100000000000000000000)
alu_b <= 0x000003fc (00000000000000000000001111111100)
alu_r :  0xbfd003fc (10111111110100000000001111111100)
R[17] <= 0xbfd003fc (10111111110100000000001111111100)

fetch instr @ 0x80000018 (10000000000000000000000000011000)
MMU[0x80000018] :  0x00000018
Mem[0x00000018] :  0x3c044000 (00111100000001000100000000000000)
001111 | 00000 | 00100 | 0100000000000000
LUI      rs: 0   rt: 4   imm:0x4000
R[ 4] <= 0x40000000 (01000000000000000000000000000000)

fetch instr @ 0x8000001c (10000000000000000000000000011100)
MMU[0x8000001c] :  0x0000001c
Mem[0x0000001c] :  0x8c840070 (10001100100001000000000001110000)
100011 | 00100 | 00100 | 0000000001110000
LW       rs: 4   rt: 4   imm:0x0070
R[ 4] :  0x40000000 (01000000000000000000000000000000)
alu_a <= 0x40000000 (01000000000000000000000000000000)
alu_b <= 0x00000070 (00000000000000000000000001110000)
alu_r :  0x40000070 (01000000000000000000000001110000)
Exception (ExcCode:2)

fetch instr @ 0x80000180 (10000000000000000000000110000000)
MMU[0x80000180] :  0x00000180
Mem[0x00000180] :  0x241a0000 (00100100000110100000000000000000)
001001 | 00000 | 11010 | 0000000000000000
ADDIU    rs: 0   rt:26   imm:0x0000
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0x00000000 (00000000000000000000000000000000)
alu_r :  0x00000000 (00000000000000000000000000000000)
R[26] <= 0x00000000 (00000000000000000000000000000000)

fetch instr @ 0x80000184 (10000000000000000000000110000100)
MMU[0x80000184] :  0x00000184
Mem[0x00000184] :  0x409a0000 (01000000100110100000000000000000)
010000 | 00100 | 11010 | 00000 | 00000000 | 000
COP0     mtc0    rt:26   Index   
R[26] :  0x00000000 (00000000000000000000000000000000)
R[Index   ] <= 0x00000000 (00000000000000000000000000000000)

fetch instr @ 0x80000188 (10000000000000000000000110001000)
MMU[0x80000188] :  0x00000188
Mem[0x00000188] :  0x241a0002 (00100100000110100000000000000010)
001001 | 00000 | 11010 | 0000000000000010
ADDIU    rs: 0   rt:26   imm:0x0002
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0x00000002 (00000000000000000000000000000010)
alu_r :  0x00000002 (00000000000000000000000000000010)
R[26] <= 0x00000002 (00000000000000000000000000000010)

fetch instr @ 0x8000018c (10000000000000000000000110001100)
MMU[0x8000018c] :  0x0000018c
Mem[0x0000018c] :  0x409a1000 (01000000100110100001000000000000)
010000 | 00100 | 11010 | 00010 | 00000000 | 000
COP0     mtc0    rt:26   EntryLo0
R[26] :  0x00000002 (00000000000000000000000000000010)
R[EntryLo0] <= 0x00000002 (00000000000000000000000000000010)

fetch instr @ 0x80000190 (10000000000000000000000110010000)
MMU[0x80000190] :  0x00000190
Mem[0x00000190] :  0x3c1a4000 (00111100000110100100000000000000)
001111 | 00000 | 11010 | 0100000000000000
LUI      rs: 0   rt:26   imm:0x4000
R[26] <= 0x40000000 (01000000000000000000000000000000)

fetch instr @ 0x80000194 (10000000000000000000000110010100)
MMU[0x80000194] :  0x00000194
Mem[0x00000194] :  0x409a5000 (01000000100110100101000000000000)
010000 | 00100 | 11010 | 01010 | 00000000 | 000
COP0     mtc0    rt:26   EntryHi 
R[26] :  0x40000000 (01000000000000000000000000000000)
R[EntryHi ] <= 0x40000000 (01000000000000000000000000000000)

fetch instr @ 0x80000198 (10000000000000000000000110011000)
MMU[0x80000198] :  0x00000198
Mem[0x00000198] :  0x42000002 (01000010000000000000000000000010)
010000 | 1 | 0000000000000000000 | 000010
COP0    CO                     0   TLBWI

fetch instr @ 0x8000019c (10000000000000000000000110011100)
MMU[0x8000019c] :  0x0000019c
Mem[0x0000019c] :  0x42000018 (01000010000000000000000000011000)
010000 | 1 | 0000000000000000000 | 011000
COP0    CO                     0   ERET
Exception Return

fetch instr @ 0x8000001c (10000000000000000000000000011100)
MMU[0x8000001c] :  0x0000001c
Mem[0x0000001c] :  0x8c840070 (10001100100001000000000001110000)
100011 | 00100 | 00100 | 0000000001110000
LW       rs: 4   rt: 4   imm:0x0070
R[ 4] :  0x40000000 (01000000000000000000000000000000)
alu_a <= 0x40000000 (01000000000000000000000000000000)
alu_b <= 0x00000070 (00000000000000000000000001110000)
alu_r :  0x40000070 (01000000000000000000000001110000)
Mem[0x00000070] :  0x00000000 (00000000000000000000000000000000)
R[ 4] <= 0x00000000 (00000000000000000000000000000000)

fetch instr @ 0x80000020 (10000000000000000000000000100000)
MMU[0x80000020] :  0x00000020
Mem[0x00000020] :  0x0c000016 (00001100000000000000000000010110)
000011 | 00000000000000000000010110
JAL      instr_index:0x0000016
R[31] <= 0x80000028 (10000000000000000000000000101000)

fetch instr @ 0x80000058 (10000000000000000000000001011000)
MMU[0x80000058] :  0x00000058
Mem[0x00000058] :  0x24090001 (00100100000010010000000000000001)
001001 | 00000 | 01001 | 0000000000000001
ADDIU    rs: 0   rt: 9   imm:0x0001
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0x00000001 (00000000000000000000000000000001)
alu_r :  0x00000001 (00000000000000000000000000000001)
R[ 9] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x8000005c (10000000000000000000000001011100)
MMU[0x8000005c] :  0x0000005c
Mem[0x0000005c] :  0x8e280000 (10001110001010000000000000000000)
100011 | 10001 | 01000 | 0000000000000000
LW       rs:17   rt: 8   imm:0x0000
R[17] :  0xbfd003fc (10111111110100000000001111111100)
alu_a <= 0xbfd003fc (10111111110100000000001111111100)
alu_b <= 0x00000000 (00000000000000000000000000000000)
alu_r :  0xbfd003fc (10111111110100000000001111111100)
MMU[0xbfd003fc] :  0x1fd003fc
Mem[0x1fd003fc] :  0x00000001 (00000000000000000000000000000001)
R[ 8] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x80000060 (10000000000000000000000001100000)
MMU[0x80000060] :  0x00000060
Mem[0x00000060] :  0x31080001 (00110001000010000000000000000001)
001100 | 01000 | 01000 | 0000000000000001
ANDI     rs: 8   rt: 8   imm:0x0001
R[ 8] :  0x00000001 (00000000000000000000000000000001)
alu_a <= 0x00000001 (00000000000000000000000000000001)
alu_b <= 0x00000001 (00000000000000000000000000000001)
alu_r :  0x00000001 (00000000000000000000000000000001)
R[ 8] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x80000064 (10000000000000000000000001100100)
MMU[0x80000064] :  0x00000064
Mem[0x00000064] :  0x1509fffd (00010101000010011111111111111101)
000101 | 01000 | 01001 | 1111111111111101
BNE      rs: 8   rt: 9   imm:0xfffd
R[ 8] :  0x00000001 (00000000000000000000000000000001)
R[ 9] :  0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x80000068 (10000000000000000000000001101000)
MMU[0x80000068] :  0x00000068
Mem[0x00000068] :  0xae040000 (10101110000001000000000000000000)
101011 | 10000 | 00100 | 0000000000000000
SW       rs:16   rt: 4   imm:0x0000
R[16] :  0xbfd003f8 (10111111110100000000001111111000)
R[ 4] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0xbfd003f8 (10111111110100000000001111111000)
alu_b <= 0x00000000 (00000000000000000000000000000000)
alu_r :  0xbfd003f8 (10111111110100000000001111111000)
MMU[0xbfd003f8] :  0x1fd003f8
Mem[0x1fd003f8] <= 0x00 (00000000)

fetch instr @ 0x8000006c (10000000000000000000000001101100)
MMU[0x8000006c] :  0x0000006c
Mem[0x0000006c] :  0x03e00008 (00000011111000000000000000001000)
000000 | 11111 | 00000 | 00000 | 00000 | 001000
SPECIAL  rs:31   rt: 0   rd: 0   sa: 0   JR     
R[31] :  0x80000028 (10000000000000000000000000101000)

fetch instr @ 0x80000028 (10000000000000000000000000101000)
MMU[0x80000028] :  0x00000028
Mem[0x00000028] :  0x0000000d (00000000000000000000000000001101)
000000 | 00000 | 00000 | 00000 | 00000 | 001101
SPECIAL  rs: 0   rt: 0   rd: 0   sa: 0   BREAK  
halt
